hyperbahn/lib/larch/larch.js

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var assert = require('assert');
+var collectParallel = require('collect-parallel/array');
+
+var Record = require('./record');
+
+module.exports = Larch;
+
+function Larch(options) {
+    if (!(this instanceof Larch)) {
+        return new Larch(options);
+    }
+
+    var self = this;
+
+    self.backends = options.backends;
+    assert(Array.isArray(self.backends), 'options.backends must be array');
+
+    if (self.backends.length === 1) {
+        self.log = self.logSingleBackend;
+    } else {
+        self.log = self.logMultiBackend;
+    }
+}
+
+Larch.prototype.logSingleBackend =
+function logSingleBackend(level, msg, meta, cb) {
+    var self = this;
+
+    var record = new Record(level, msg, meta, null);
+
+    self.backends[0].log(record, cb);
+};
+
+Larch.prototype.logMultiBackend =
+function logMultiBackend(level, msg, meta, cb) {
+    var self = this;
+
+    var record = new Record(level, msg, meta, null);
+
+    var i;
+    for (i = 0; i < self.backends.length; i++) {
+        self.backends[i].log(record, backendDone);
+    }
+
+    var done = 0;
+    function backendDone() {
+        done++;
+
+        if (done >= self.backends.length) {
+            cb();
+        }
+    }
+};
+
+Larch.prototype.bootstrap = function bootstrap(cb) {
+    var self = this;
+
+    collectParallel(self.backends, bootstrapBackend, cb);
```

Comment:
collect parallel returns (null, Array<{ err, value }>). This is probably not what you wanted.